src/github.com/couchbase/sync_gateway/db/channel_cache.go

```diff
@@ -257,3 +307,140 @@ func (c *channelCache) prependChanges(changes LogEntries, changesValidFrom uint6
 		return 0
 	}
 }
+
+type lateLogEntry struct {
+	logEntry      *LogEntry
+	arrived       time.Time
+	listenerCount uint64
+	listenerLock  sync.RWMutex
+}
+
+func (l *lateLogEntry) addListener() {
+	l.listenerLock.Lock()
+	l.listenerCount++
+	l.listenerLock.Unlock()
+}
+
+func (l *lateLogEntry) removeListener() {
+	l.listenerLock.Lock()
+	l.listenerCount--
+	l.listenerLock.Unlock()
+}
+
+func (l *lateLogEntry) getListenerCount() uint64 {
+	l.listenerLock.RLock()
+	defer l.listenerLock.RUnlock()
+	return l.listenerCount
+}
+
+// Initialize the late-arriving log queue with a zero entry, used to track listeners.  This is needed
+// to support purging later entries once everyone has seen them.
+func (c *channelCache) initializeLateLogs() {
+	log := &LogEntry{Sequence: 0}
+	lateEntry := &lateLogEntry{
+		logEntry:      log,
+		listenerCount: 0,
+	}
+	c.lateLogs = append(c.lateLogs, lateEntry)
+}
+
+// Retrieve late-arriving sequences that have arrived since the previous sequence.  Retrieves set of sequences, and the last
+// sequence number in the list.  Note that lateLogs is sorted by arrival on Tap, not sequence number.
+func (c *channelCache) GetLateSequencesSince(sinceSequence uint64) (entries []*LogEntry, lastSequence uint64, err error) {
+
+	c.lateLogLock.RLock()
+	defer c.lateLogLock.RUnlock()
+
+	// If we don't have any new late sequences, return empty
+	if sinceSequence == c.lastLateSequence {
+		return nil, sinceSequence, nil
+	}
+
+	// Return everything in lateLogs after sinceSequence (the caller's last seen sequence).
+	var previousEntry *lateLogEntry
+	previousFound := false
+
+	for index, lateLog := range c.lateLogs {
+		if previousFound {
+			entries = append(entries, lateLog.logEntry)
+		} else {
+			if lateLog.logEntry.Sequence == sinceSequence {
+				// Found the previous sequence.
+				entries = make([]*LogEntry, 0, len(c.lateLogs)-index)
+				previousEntry = lateLog
+				previousFound = true
+			}
+		}
+	}
+
+	// If we didn't find the previous sequence, the consumer may be missing sequences and must rollback
+	// to their last safe sequence
+	if !previousFound {
+		err = errors.New("Missing previous sequence - rollback required")
+		return
+	}
+
+	// Remove listener from the sinceSequence entry, and add a listener to the latest
+	previousEntry.removeListener()
+	c.lateLogs[len(c.lateLogs)-1].addListener()
+	lastSequence = c.lateLogs[len(c.lateLogs)-1].logEntry.Sequence
+
+	return entries, lastSequence, nil
+
+}
+
+// Called on first call to the channel during changes processing, to get starting point for
+// subsequent checks for late arriving sequences.
+func (c *channelCache) InitLateSequenceClient() uint64 {
+	log := c.lateLogs[len(c.lateLogs)-1]
```

Comment:
instead of calling `c.lateLogs[len(c.lateLogs)-1]` in several places, maybe it makes sense to add a method: `c.mostRecentLateLog()`